File: JAVA/28_Tree/BOJ_04_1991_트리의순회.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class node {
	char data;
	node left, right;

	public node(char data) {
		this.data = data;
	}
}

public class Main {

	static node[] tree;
	static StringBuilder sb;

	public static void preorder(char data) {
		sb.append(tree[data - 'A'].data);
		if (tree[data - 'A'].left != null) {
			preorder(tree[data - 'A'].left.data);
		}
		if (tree[data - 'A'].right != null) {
			preorder(tree[data - 'A'].right.data);
		}
	}

	public static void inorder(char data) {
		if (tree[data - 'A'].left != null) {
			inorder(tree[data - 'A'].left.data);
		}
		sb.append(tree[data - 'A'].data);
		if (tree[data - 'A'].right != null) {
			inorder(tree[data - 'A'].right.data);
		}
	}

	public static void postorder(char data) {
		if (tree[data - 'A'].left != null) {
			postorder(tree[data - 'A'].left.data);
		}
		if (tree[data - 'A'].right != null) {
			postorder(tree[data - 'A'].right.data);
		}
		sb.append(tree[data - 'A'].data);
	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		tree = new node[n];
		for (int i = 0; i < n; i++) {
			tree[i] = new node((char) ('A' + i));
		}
		for (int i = 0; i < n; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine(), " ");
			char root = st.nextToken().charAt(0);
			char left = st.nextToken().charAt(0);
			char right = st.nextToken().charAt(0);
			if (left != '.') {
				tree[root - 'A'].left = tree[left - 'A'];
			}
			if (right != '.') {
				tree[root - 'A'].right = tree[right - 'A'];
			}
		}
		sb = new StringBuilder();
		preorder('A');
		sb.append("\n");
		inorder('A');
		sb.append("\n");
		postorder('A');
		System.out.print(sb);
	}
}

File: C/28_Tree/BOJ_04_1991_트리의순회.c
#include <stdio.h>
#include <stdlib.h>
typedef struct node {
	char data;
	struct node* left;
	struct node* right;
} node;
node* tree[26];
void preorder(char data) {
	printf("%c", tree[data - 'A']->data);
	if (tree[data - 'A']->left != NULL) {
		preorder(tree[data - 'A']->left->data);
	}
	if (tree[data - 'A']->right != NULL) {
		preorder(tree[data - 'A']->right->data);
	}
}
void inorder(char data) {
	if (tree[data - 'A']->left != NULL) {
		inorder(tree[data - 'A']->left->data);
	}
	printf("%c", tree[data - 'A']->data);
	if (tree[data - 'A']->right != NULL) {
		inorder(tree[data - 'A']->right->data);
	}
}
void postorder(char data) {
	if (tree[data - 'A']->left != NULL) {
		postorder(tree[data - 'A']->left->data);
	}
	if (tree[data - 'A']->right != NULL) {
		postorder(tree[data - 'A']->right->data);
	}
	printf("%c", tree[data - 'A']->data);
}
main() {
	int n;
	scanf("%d", &n);
	for (int i = 0; i < n; i++) {
		tree[i] = (node*)malloc(sizeof(node));
		tree[i]->data = 'A' + i;
		tree[i]->left = NULL;
		tree[i]->right = NULL;
	}
	for (int i = 0; i < n; i++) {
		getchar();
		char root, left, right;
		scanf("%c %c %c", &root, &left, &right);
		if (left != '.') {
			tree[root - 'A']->left = tree[left - 'A'];
		}
		if (right != '.') {
			tree[root - 'A']->right = tree[right - 'A'];
		}
	}
	preorder('A');
	printf("\n");
	inorder('A');
	printf("\n");
	postorder('A');
	printf("\n");
}